### ISSUE TITLE
Automatically inserting `end` in Ruby

### ISSUE THREAD

**Issue by janko** [zed-industries/zed#12383]
### Check for existing issues

- [X] Completed

### Misc notes

From vim-ruby, I'm used to `end` being automatically inserted when I type `def something<Enter>` or `foo do<Enter>` or `case<Enter>`. Could this functionality be added to the Ruby extension?

**Comment by zoedsoupe** [comment]
that works too with Elixir (:

```elixir
defmodule Foo do
  def bar do
    # body...
  end
end
```

**Comment by MrSubidubi** [comment]
Thanks for the suggestion! 

That definitely sounds reasonable, although I am not sure whether this could/should be provided by the extension. As per https://github.com/zed-extensions/ruby/issues/134 it does not seem to be the case. I hope that's fine for me to ping you, @vitallium , may I ask whether you have some more context on this?

**Comment by janko** [comment]
This appears to be a duplicate of https://github.com/zed-extensions/ruby/issues/72, but I think it's a Zed issue and not an extension issue. Ruby LSP implements the automatic `end` insertion, but Zed doesn't send the `document/onTypeFormatting` request when pressing <kbd>enter</kbd> at the end of `def foo`.

**Comment by MrSubidubi** [comment]
Well, we should be handling this, as per https://github.com/zed-industries/zed/blob/f46ad0c3e9bd477e2d7db1b63fe8be3b8d695a7c/crates/project/src/lsp_store.rs#L3752. Can you check what the Ruby LSP returns in its server info for these capabilities? You can find them in `dev: open language server logs` -> `Server Info`

**Comment by janko** [comment]
Yes, Ruby LSP declares this capability, and it's [documented](https://shopify.github.io/ruby-lsp/#on-type-formatting) to work in VSCode. Here is the relevant part from server info:

```json
"documentOnTypeFormattingProvider": {
  "firstTriggerCharacter": "{",
  "moreTriggerCharacter": [
    "\n",
    "|",
    "d"
  ]
}
```

The Ruby Zed extension disables it by default because of a [different issue](https://github.com/zed-extensions/ruby/issues/38), but it can be re-enabled:

```json
{
  "lsp": {
    "ruby-lsp": {
      "initialization_options": {
        "enabledFeatures": {
          "onTypeFormatting": true
        }
      }
    }
  }
}
```

Zed sends the `document/onTypeFormatting` request when typing `|` (which proves it's working with Ruby LSP), but not when typing <kbd>enter</kbd>.
